Expand pool.name in semantic conventions (#3050)
open-telemetry/opentelemetry-specification: specification/metrics/semantic_conventions/database-metrics.md

| Name        | Type   | Description                                                                  | Examples       | Requirement Level |
|-------------|--------|------------------------------------------------------------------------------|----------------|-------------------|

> | `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the DB we are working on, then the name should be bui as follows: `host: <host> port: <port> database: <database name> user: <user>`. | `myDataSource`, `host: localhost port: 3306 database: db_test user: root`  | Required          |

Reviewer: ```suggestion
| `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the monitored database, then the name should be as follows: `host: <host> port: <port> database: <database name> user: <user>`. | `myDataSource`, `host: localhost port: 3306 database: db_test user: root`  | Required          |
```

> | `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the DB we are working on, then the name should be bui as follows: `host: <host> port: <port> database: <database name> user: <user>`. | `myDataSource`, `host: localhost port: 3306 database: db_test user: root`  | Required          |

Author: No problem. I'll change it.

> | `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the DB we are working on, then the name should be bui as follows: `host: <host> port: <port> database: <database name> user: <user>`. | `myDataSource`, `host: localhost port: 3306 database: db_test user: root`  | Required          |

Author: I added it in another PR: https://github.com/open-telemetry/opentelemetry-specification/pull/3053 sorry for the confusion.

> | `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the DB we are working on, then the name should be bui as follows: `host: <host> port: <port> database: <database name> user: <user>`. | `myDataSource`, `host: localhost port: 3306 database: db_test user: root`  | Required          |

Reviewer: Thank you, but could you please add the change in this PR?
This way it's ensured to stay consistent as it can only merge atomically.

> | `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the DB we are working on, then the name should be bui as follows: `host: <host> port: <port> database: <database name> user: <user>`. | `myDataSource`, `host: localhost port: 3306 database: db_test user: root`  | Required          |

Author: done

> | `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the monitored database, then the name should be as follows: `host: '<host>', port: <port>, database: '<database name>', user: '<user>'`. | `myDataSource`, `host: localhost port: 3306 database: db_test user: root`  | Required          |

Reviewer 2: Nit: consider adding the database connection string as a possible option:
```suggestion
| `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the monitored database, consider using the connection string (with password removed). Alternatively, you can use the following pattern: `host: '<host>', port: <port>, database: '<database name>', user: '<user>'`. | `myDataSource`, `host: localhost, port: 3306, database: db_test, user: root`  | Required          |
```

The idea behind the `pool.name` attribute was to make it as easy as possible for the owner of the instrumented app to find out which metric measurements are produced by which connection pools in their app; if the connection pool implementation does not name the pools, usually (if available) the connection string is a good option for differentiating between different connection pools (because, if you have >1 DB connection pools in your app, usually they connect to different DBs/schemas/users).

Reviewer 3: Why isn't pool name just optional?

Reviewer 2: You can have multiple connection pools in your application; and if you do not provide the name, how will you distinguish between them? 
Also, almost all instruments used here are up down counters, if you were to provide a measurement without a name it'd break the "additive" property that counters exhibit (measurement without a name != sum of measurements).

Reviewer 4: Connection pools aren't required to use a database, though, are they?  I know that they're best practice but it is certainly possible to use, and instrument the use of, a database without connection pooling.

Reviewer: The connection string could be an option. We also have it as a dedicated attribute for traces already (and also just recommend to strip credentials there): https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#connection-level-attributes

Author: You mean that instead of "building" the pool name as suggested here, we will use the already existing parameter DB_CONNECTION_STRING to identify the pool name? 

If yes, then I should change the definition of pool.name above to something like that: "...In case the pool has no name in the monitored database, then the name should be the same as 'db.connection_string' defined [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#connection-level-attributes)". agreed?

Reviewer: I'm not sure if the entire (redacted) connection string is suitable for metrics with regard to cardinality.
@open-telemetry/specs-metrics-approvers any opinions?

Reviewer: As for omitting `pool.name` in favor of `db.connection_string` in certain cases, this might be problematic for metrics consumers, as they'd have to consider both attributes for querying and aggregating/splitting them.

If we want to go with the connection string fallback, defining it to be set as `pool.name` might be better than introducing a separate `db.connection_string` for metrics - currently we only have it for traces.

Author: yes. that was exactly the purpose (at least as I understood it). in case we don't have the name of the pool, we will do: `setAttribure(pool.name, <the values of db.connection_string>)` - meaning that we will set the `pool.name` to be the value of the connection_string.  @mateuszrzeszutek  what do you think?

Reviewer 2: > If we want to go with the connection string fallback, defining it to be set as `pool.name` might be better than introducing a separate `db.connection_string` for metrics - currently we only have it for traces.

👍 
`pool.name` values have to be different in different connection pools; and ideally it should be easy to find a particular pool by its name in the application code. I think that the connection string somewhat fulfills these conditions; if a particular connection pool implementation does not provide a name the connection string should work as a fallback.

Author: I agree with you. I committed another change. please review it when you have time. thanks.

> | `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the monitored database, then the name should be as follows: `host: '<host>', port: <port>, database: '<database name>', user: '<user>'`. | `myDataSource`, `host: localhost port: 3306 database: db_test user: root`  | Required          |

Reviewer: ```suggestion
| `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the monitored database, then the name should be as follows: `host: '<host>', port: <port>, database: '<database name>', user: '<user>'`. | `myDataSource`, `host: localhost, port: 3306, database: 'db_test', user: 'root'`  | Required          |
```

> | `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the pool has no name in the monitored database, then the [db.connection_string](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#connection-level-attributes) should be used. | `myDataSource` | Required          |

Reviewer 2: Nit
```suggestion
| `pool.name` | string | The name of the connection pool; unique within the instrumented application. In case the connection pool implementation does not provide a name, then the [db.connection_string](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#connection-level-attributes) should be used. | `myDataSource` | Required          |
```
